### DISCUSSION TITLE
Encoding problems with old systems

### DISCUSSION THREAD

**Discussion by AcclaroDev** [gitextensions/gitextensions#11080] [Q&A]
Hello everyone,
First of all, I love Git Extentions. I searched around for git UIs some years ago and Git Extentions pretty much gives you most options, without needing to remember command line commands. It replaced all my other git UIs, like when I first used the thing that came with Visual Studio. It was a good starting point, but it was just lacking features everywhere.

So anyway, since we are trying to use Git Extensions also for our vintage projects, there is some weird behaviour in the diff window.
1.	I’m not sure what encoding they are exactly using, but winmerge  recognizes it as “ISO-10646-UCS-2 BOM” and kdiff3 sees it as “UTF16-LE”.
winmerge:
![image](https://github.com/gitextensions/gitextensions/assets/138010920/c035d06f-a318-4e53-a511-c2ca2df42a39)
kdiff:
![image](https://github.com/gitextensions/gitextensions/assets/138010920/f6b58421-4f56-4f1f-af11-451e3e09fa98)

I can’t really find those in the git extensions configure window “Git->Config->Files content encoding Configure” and the diff looks weird and the line numbers are wrong.
![image](https://github.com/gitextensions/gitextensions/assets/138010920/37b4434d-1fb3-4f19-a07f-7c6649ffb7a2)

2.	Changing the encoding of a file shows in the diff extensions like it deletes the whole content and inserts it with the new one. This might be intentional, but I would prefer if there were an option so it looks a bit like in the other diff tools.
![image](https://github.com/gitextensions/gitextensions/assets/138010920/551b524c-cf26-423d-a74f-91a73fdc1127)

Are there options I’m missing? Is there any chance that this will be changed / added? I wasn't sure if this is a bug or a feature request. I thought maybe I just ask here first.

**Comment by RussKie** [comment]
Encodings are hard :)

You can configure encodings in this dialog, these encodings are those that available in your OS:
![image](https://github.com/gitextensions/gitextensions/assets/4403806/771fdb2f-9582-40cf-ac34-45a4d59c2f60)

**Reply by AcclaroDev** [reply]
Hey!

That's the exact window I meant, when I said "I can’t really find those in the git extensions configure window “Git->Config->Files content encoding Configure”"

Again: winmerge tells me they are encoded in "ISO-10646-UCS-2 BOM”, kdiff tells me it's “UTF16-LE”. Neither of them are in that list. There is no ISO-10646, no UCS-2 and no UTF-16.

Do you know what the name would be in that list?

**Reply by gerhardol** [reply]
I believe the names comes from.net.
The encoding in the editor is what git reports, not detected from the file.

**Reply by vbjay** [reply]
![Screenshot_20230702-034527_Chrome.jpg](https://github.com/gitextensions/gitextensions/assets/3629489/d0f65f62-37ed-4eb8-a52f-346274b759f7)

https://learn.microsoft.com/en-us/dotnet/api/system.text.encodinginfo.name

**Reply by vbjay** [reply]
UCS. https://www.ibm.com/docs/en/i/7.4?topic=unicode-ucs-2-its-relationship-utf-16. Universal Coded Character Set or Unicode 

UTF -# Unicode Transformation Format - bits per char

BOM byte order mark. Encodings with BOM will inject their sequence of bytes to denote that the following bytes are x encoding bytes. 

ISO-10646 is the unicode standard


ISO-10646-UCS-2 BOM

https://learn.microsoft.com/en-us/dotnet/api/system.text.unicodeencoding.-ctor#system-text-unicodeencoding-ctor(system-boolean-system-boolean-system-boolean) means that the bytes have a BOM and the encoding constructor would pass true to the byteOrderMark parameter. 


So basically the System.Text.Encoding.Unicode property is UTF-16 LE

.net itself internally stores strings as this encoding in memory https://stackoverflow.com/questions/63767503/are-strings-always-little-endian-unicode
